File: Competitive Programming/Linked List/Split a Circular Linked List into two halves.py
'''https://practice.geeksforgeeks.org/problems/split-a-circular-linked-list-into-two-halves/1
Split a Circular Linked List into two halves 
Easy Accuracy: 50.84% Submissions: 28677 Points: 2
Given a Cirular Linked List of size N, split it into two halves circular lists. If there are odd number of nodes in the given circular linked list then out of the resulting two halved lists, first list should have one node more than the second list. The resultant lists should also be circular lists and not linear lists.

 

Example 1:

Input:
Circular LinkedList: 1->5->7
Output:
1 5
7
 

Example 2:

Input:
Circular LinkedList: 2->6->1->5
Output:
2 6
1 5

Your Task:
Your task is to complete the given function splitList(), which takes 3 input parameters: The address of the head of the linked list, addresses of the head of the first and second halved resultant lists and Set the head1_ref and head2_ref to the first resultant list and second resultant list respectively.


Expected Time Complexity: O(N)
Expected Auxilliary Space: O(1)


Constraints:
1 <= N <= 100'''


# User function Template for python3

'''
class Node:
    def __init__(self):
        self.data = None
        self.next = None
'''


class Solution:
    def splitList(self, head, head1, head2):
        # code here
        if head is None:
            return(None)
        slow = head
        fast = head

        while fast.next != head and fast.next.next != head:
            slow = slow.next
            fast = fast.next.next

        if fast.next != head:
            fast = fast.next

        head1 = head
        head2 = slow.next
        # make first half circular
        slow.next = head1

        # make the second half circular
        fast.next = head2
        # this is to emulate pass by reference in python please don't delete below line.
        return head1, head2

# {
#  Driver Code Starts
# Initial Template for Python 3


class Node:
    def __init__(self):
        self.data = None
        self.next = None


def printCircularLinkedList(head):
    tmp = head
    while True:
        print(tmp.data, end=" ")
        tmp = tmp.next
        if tmp == head:
            break
    print()


if __name__ == "__main__":
    for i in range(int(input())):
        head = Node()
        head1 = Node()
        head2 = Node()
        tmp = head
        n = int(input())
        for i in input().split():
            tmp.next = Node()
            tmp = tmp.next
            tmp.data = int(i)
        head = head.next
        tmp.next = head
        obj = Solution()
        head1, head2 = obj.splitList(head, head1, head2)
        printCircularLinkedList(head1)
        printCircularLinkedList(head2)


# } Driver Code Ends

File: competitive-programming/Linked List/Split a Circular Linked List into two halves.cpp
/*
https://practice.geeksforgeeks.org/problems/split-a-circular-linked-list-into-two-halves/1
Split a Circular Linked List into two halves 
Easy Accuracy: 50.84% Submissions: 28677 Points: 2
Given a Cirular Linked List of size N, split it into two halves circular lists. If there are odd number of nodes in the given circular linked list then out of the resulting two halved lists, first list should have one node more than the second list. The resultant lists should also be circular lists and not linear lists.

 

Example 1:

Input:
Circular LinkedList: 1->5->7
Output:
1 5
7
 

Example 2:

Input:
Circular LinkedList: 2->6->1->5
Output:
2 6
1 5

Your Task:
Your task is to complete the given function splitList(), which takes 3 input parameters: The address of the head of the linked list, addresses of the head of the first and second halved resultant lists and Set the head1_ref and head2_ref to the first resultant list and second resultant list respectively.


Expected Time Complexity: O(N)
Expected Auxilliary Space: O(1)


Constraints:
1 <= N <= 100
*/

// { Driver Code Starts
/* Program to split a circular linked list into two halves */

#include <stdio.h>
#include <stdlib.h>

/* structure for a Node */
struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

void splitList(struct Node *head, struct Node **head1_ref, struct Node **head2_ref);
struct Node *newNode(int key)
{
    struct Node *temp = new Node(key);

    return temp;
}
/* Function to split a list (starting with head) into two lists.
   head1_ref and head2_ref are references to head Nodes of 
    the two resultant linked lists */

void printList(struct Node *head)
{
    struct Node *temp = head;
    if (head != NULL)
    {
        do
        {
            printf("%d ", temp->data);
            temp = temp->next;
        } while (temp != head);
        printf("\n");
    }
}

/* Driver program to test above functions */
int main()
{

    int t, n, i, x;
    scanf("%d", &t);
    while (t--)
    {
        struct Node *temp, *head = NULL;
        struct Node *head1 = NULL;
        struct Node *head2 = NULL;
        scanf("%d", &n);
        scanf("%d", &x);
        head = new Node(x);
        temp = head;
        for (i = 0; i < n - 1; i++)
        {
            scanf("%d", &x);

            temp->next = new Node(x);

            temp = temp->next;
        }

        temp->next = head;

        splitList(head, &head1, &head2);

        // printf("\nFirst Circular Linked List");
        printList(head1);

        // printf("\nSecond Circular Linked List");
        printList(head2);
    }
    return 0;
}
// } Driver Code Ends

/* The structure of linked list is the following
struct Node
{
  int data;
  struct Node *next;
  
  Node(int x){
      data = x;
      next = NULL;
  }
};
*/

//  function which splits the circular linked list.  head is pointer
// to head Node of given lined list.  head1_ref1 and *head_ref2
// are pointers to head pointers of resultant two halves.

void splitList(Node *head, Node **head1_ref, Node **head2_ref)
{
    Node *slow = head, *fast = head;

    while (fast->next != head && fast->next->next != head)
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    *head1_ref = head;
    *head2_ref = slow->next;

    Node *temp = slow->next;

    slow->next = head;

    Node *curr = *head2_ref;

    while (curr->next != head)
    {
        curr = curr->next;
    }

    curr->next = *head2_ref;
}